### ISSUE TITLE
Prometheus metrics reporting on old service IPs following service restart

### ISSUE THREAD

**Issue by richardjq** [containous/traefik#9836]
### Welcome!

- [X] Yes, I've searched similar issues on [GitHub](https://github.com/traefik/traefik/issues) and didn't find any.
- [X] Yes, I've searched similar issues on the [Traefik community forum](https://community.traefik.io) and didn't find any.

### What did you do?

I am upgrading Traefik from `2.5` to `2.9`. Following the update I have noticed that if Traefik is running and a service restarts, prometheus metric traefik_service_server_up will have metrics for both the old and new IP addresses.

The service is updated, but metrics shows both the old and the new IP addresses. This makes it difficult to alert on service vailability.

Note in the output below that `10.0.76.28` and `10.0.76.29` are the IP addresses of the service prior to restarting it (via `docker stack rm my_service; docker stack deploy...`) The other two IP addresses (`76.235` and `76.236` are the new IP addresses following the service restart.

```
/ # wget -q -O - http://localhost:8080/metrics|grep my_service
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.235:8080"} 1
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.236:8080"} 1
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.28:8080"} 0
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.29:8080"} 0
```

I expect that following a service restart, Traefik should no longer report on metrics for the old IP addresses.

```
/ # wget -q -O - http://localhost:8080/metrics|grep my_service
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.235:8080"} 1
traefik_service_server_up{service="my_service@docker",url="http://10.0.76.236:8080"} 1
```

Fetching the service configuration from the api, I see that it is as expected (only have the new IP addresses).
The json below shows what is returned from  `wget -q -O -  http://localhost:8080/api/http/services/my_service@docker`

```json
{
	"loadBalancer": {
		"sticky": {
			"cookie": {
				"name": "X-Route-my_service",
				"secure": true
			}
		},
		"servers": [{
			"url": "http://10.0.76.236:8080"
		}, {
			"url": "http://10.0.76.235:8080"
		}],
		"healthCheck": {
			"scheme": "http",
			"path": "/myService/?monitor=traefik-healthcheck",
			"port": 8080,
			"interval": "30s",
			"followRedirects": false
		},
		"passHostHeader": true
	},
	"status": "enabled",
	"usedBy": ["my_service@docker"],
	"serverStatus": {
		"http://10.0.76.235:8080": "UP",
		"http://10.0.76.236:8080": "UP"
	},
	"name": "my_service@docker",
	"provider": "docker",
	"type": "loadbalancer"
}
```

### What did you see instead?

I expect to see only the metrics for the new service IPs when querying `/metrics`. What I actually see is metrics for both the old IP addresses and the new IP addresses.

### What version of Traefik are you using?

Version:      2.9.10
Codename:     banon
Go version:   go1.20.3
Built:        2023-04-06T16:15:08Z
OS/Arch:      linux/amd64

### What is your environment & configuration?

Excerpt from docker compose file:
```
    command:
      - "--providers.docker=true"
      - "--providers.docker.endpoint=unix:///var/run/docker.sock"
      - "--providers.docker.swarmMode=true"
      - "--providers.docker.exposedbydefault=false"
      - "--providers.docker.constraints=Label(`traefik.tier`,`pprd`)"
      - "--providers.docker.network=pprd_apps_es_vt_edu_encrypted"
      - "--entryPoints.http.address=:80"
      - "--entryPoints.http.forwardedHeaders.insecure"
      - "--metrics.prometheus=true"
      - "--api.dashboard=true"
      - "--api.insecure=true"
      - "--accesslog.format=json"
      - "--log.format=json"
      - "--ping"
```

**Comment by jasonrclemons** [comment]
same issue. seeking answers.

**Comment by rtribotte** [comment]
Hello @richardjq @jasonrclemons,

Thanks for your interest in Traefik!

I tried to reproduce the issue with no success.
It is possible that the deletion of the metric for the old service URLs takes some time, thus eventually it would be deleted.

> I expect that following a service restart, Traefik should no longer report on metrics for the old IP addresses.

Are you able to consistently see the old service metrics reported over multiple "fetches" of the metrics?

**Comment by richardjq** [comment]
We run a cluster of 3 Traefiks and we restart services daily. At least one of these 3 instances experience this problem every day. However, not all 3 of them exhibit the problem. So it is intermittent.

We do continue to see this behavior for hours after the service restarts have occurred.

**Comment by richardjq** [comment]
They do not. We have waited as long as 3 hours and they are still there.

**Comment by mpl** [comment]
> They do not. We have waited as long as 3 hours and they are still there.

Yes, I had understood that the problem persists for a few hours, but I'm wondering about anything longer. Have you been able to observe it for days, or is there anything you're doing that prevents you from testing such a long observation?
I'm not expecting anything in particular, but I'm just curious about it.

**Comment by richardjq** [comment]
I have not tested for a longer period of time. The metric itself is not very useful if it provides inaccurate information for hours of the day. 

I am currently working to try to pin down which specific version introduced the problem. Each test takes a day, so it's slow going.

**Comment by richardjq** [comment]
Additional testing confirms that this but was introduced in v2.8.1.

**Comment by rtribotte** [comment]
Thanks for your feedback @richardjq.

We identified this as a (confirmed) bug and we are working on it.

**Comment by francescor** [comment]
we need to follow this, too, thanks

**Comment by francescor** [comment]
> Hey [@francescor](https://github.com/francescor),
> 
> Thank you for reaching out. What version are you using? Is it v3 or still v2?

v.2 (latest)

**Comment by nmengin** [comment]
Hello @francescor,

Thank you for the feedback.

Could you confirm if the issue is still present in Traefik v3?
Indeed, Traefik v2.11 will now be updated to bring [security fixes](https://doc.traefik.io/traefik/deprecation/releases/#versions), the other bugs are only fixed in v3.

**Comment by francescor** [comment]
> Could you confirm if the issue is still present in Traefik v3?

we won't be able to testing v3 earlier then 3 months, sorry

**Comment by traefiker** [comment]
Hi! I'm Træfiker :robot: the bot in charge of tidying up the issues.I have to close this one because of its lack of activity :disappointed:Feel free to re-open it or join our [Community Forum](https://community.containo.us/).
